### PR TITLE
new metrics + new vignette: making new layers via subclassing

### PR DESCRIPTION
**Pull request by t-kalinowski**

### CODE REVIEW

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 1]
```diff
@@ -0,0 +1,969 @@
...
+function.
+
+```{r}
+x <- tf$ones(shape(2, 2))
+linear_layer <- Linear(4, 2)
+y <- linear_layer(x)
+print(y)
+```
+
+All the builtin layers in the keras package (e.g., `layer_dense()`) are
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:69]: very nicely explained!

Comment (skeydan) [reply vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:69]: on the other hand ... seeing this again some time later, now i think, - won't we always want this?
Do it transparently to the user?

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 2]
```diff
@@ -0,0 +1,969 @@
...
+layer_linear <- create_layer_wrapper(Linear)
+layer_linear
+# function (object, units = 32, input_dim = 32)
+# {
+#     args <- capture_args(match.call(), modifiers)
+#     create_layer(LayerClass, object, args)
+# }
+# <environment: 0x557cc3d71158>
+```
+Now `layer_linear` is a layer constructor composes nicely with `%>%`, just like the built in layers:
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:85]:
```suggestion
Now `layer_linear` as a layer constructor composes nicely with `%>%`, just like the built-in layers:
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 3]
```diff
@@ -0,0 +1,969 @@
...
+```{r}
+model <- keras_model_sequential() %>%
+  layer_linear(4, 2)
+model(k_ones(c(2, 2))) |> invisible() # force model to build
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:89]: |> invisible() # force model to build

is this for education purposes? :-))))

In that case, I get it, - but doesn't it still add complexity where none is needed?

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 4]
```diff
@@ -0,0 +1,969 @@
...
+being set as layer attributes:
+
+```{r}
+stopifnot(all.equal(
+  linear_layer$weights,
+  list(linear_layer$w, linear_layer$b)
+))
+```
+
+Note you also have access to a quicker shortcut for adding weight to a layer:
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:115]:
```suggestion
Note you also have access to a quicker shortcut for adding a weight to a layer:
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 5]
```diff
@@ -0,0 +1,969 @@
...
+the `add_weight()` method:
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:116]: i think that if the guide wasn't a translation, i'd only show this way of doing it ...

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 6]
```diff
@@ -0,0 +1,969 @@
...
+print(y)
+```
+
+## Layers can have non-trainable weights
+
+Besides trainable weights, you can add non-trainable weights to a layer as
+well. Such weights are meant not to be taken into account during
+backpropagation, when you are training the layer.
+
+Here's how to add and use a non-trainable weight:
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:153]: ok, i guess this answers what i just wrote about the `tf$Variable` way ...

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 7]
```diff
@@ -0,0 +1,969 @@
...
+Linear <- R6::R6Class(
+  classname = "Linear",
+  inherit = keras$layers$Layer,
+  public = list(
+    initialize = function(units = 32) {
+      super$initialize()
+      self$units <- units
+    },
+    build = function(input_shape) {
+      print(sys.call())
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:237]: guess this should be removed? :-))

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 8]
```diff
@@ -0,0 +1,969 @@
...
+
+  call <- function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+}
+```
+
+
+The `build()` method of your layer will automatically run the first time
+your layer instance is called. You now have a layer that's lazy and thus easier to use:
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:291]:
```suggestion
your layer instance is called. You now have a layer that can handle an arbitrary number of input features:
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 9]
```diff
@@ -0,0 +1,969 @@
...
+
+In many cases, you may not know in advance the size of your inputs, and you
+would like to lazily create weights when that value becomes known, some time
+after instantiating the layer.
+
+In the Keras API, we recommend creating layer weights in the `build(self,
+inputs_shape)` method of your layer. Like this:
+
+```{r}
+# Same as below, but using R6.
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:225]: Will you show both in the end? In this case, I'd start with the Python-ey way, because it's what you mostly use, and then say sth like: "If you use keras in a package, ..." etc

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 10]
```diff
@@ -0,0 +1,969 @@
...
+  call <- function(inputs) {
+    self$add_loss(self$rate * tf$reduce_sum(inputs))
+    inputs
+  }
+}
+# layer_activity_regularization <- create_layer_wrapper(ActivityRegularizationLayer)
+```
+
+These losses (including those created by any inner layer) can be retrieved via
+`layer.losses`. This property is reset at the start of every `__call__()` to
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd]:
```suggestion
`layer$losses`. This property is reset at the start of every `call()` to
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 11]
```diff
@@ -0,0 +1,969 @@
...
+  .mapply(if (simplify) c else list, list(...), NULL)
+}
+```
+
+```{r, eval=FALSE}
+# Instantiate an optimizer.
+optimizer <- optimizer_sgd(learning_rate = 1e-3)
+loss_fn <- loss_sparse_categorical_crossentropy(from_logits = TRUE)
+
+# while(!is.null(batch <- iter_next(dataset_iterator))) { ... } ## too compact?
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:427]: i find this easiest - the "while" way i'm referring to

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 12]
```diff
@@ -0,0 +1,969 @@
...
+model %>% compile(optimizer = "adam", loss = "mse")
+model %>% fit(k_random_uniform(c(2, 3)),
+  k_random_uniform(c(2, 3)),
+  epochs = 1, verbose = FALSE
+)
+
+# It's also possible not to pass any loss in `compile`,
+# since the model already has a loss to minimize, via the `add_loss`
+# call during the forward pass!
+model %>% compile(optimizer = "adam")
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:474]: so great this works now! Remember i needed a workaround in tfprobability, in the past...

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 13]
```diff
@@ -0,0 +1,969 @@
...
+model %>% fit(k_random_uniform(c(2, 3)),
+  k_random_uniform(c(2, 3)),
+  epochs = 1, verbose = FALSE
+)
+```
+
+## The `add_metric()` method
+
+Similarly to `add_loss()`, layers also have an `add_metric()` method
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:483]: this is a great addition!

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 14]
```diff
@@ -0,0 +1,969 @@
...
+
+    # Return the inference-time prediction tensor (for `.predict()`).
+    tf$nn$softmax(logits)
+  }
+}
+```
+
+
+
+Metrics tracked in this way are accessible via `layer.metrics`:
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:519]:
```suggestion
Metrics tracked in this way are accessible via `layer$metrics`:
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 15]
```diff
@@ -0,0 +1,969 @@
...
+evaluation loops (e.g. `fit()`) to correctly use the layer in training and
+inference.
+
+```{r}
+CustomDropout(keras$layers$Layer) %py_class% {
+  initialize <- function(rate, ...) {
+    super$initialize(...)
+    self$rate <- rate
+  }
+  call <- function(inputs, training = NULL) {
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:668]: is NULL a good default? instead of FALSE i mean

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 16]
```diff
@@ -0,0 +1,969 @@
...
+```
+
+```
+
+```
+## Putting it all together: an end-to-end example
+
+Here's what you've learned so far:
+
+- A `Layer` encapsulate a state (created in `initialize()` or `build()`) and some
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:762]:
```suggestion
- A `Layer` encapsulates a state (created in `initialize()` or `build()`), and some
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 17]
```diff
@@ -0,0 +1,969 @@
...
+
+optimizer <- optimizer_adam(learning_rate = 1e-3)
+
+vae %>% compile(optimizer, loss = loss_mean_squared_error())
+vae %>% fit(x_train, x_train, epochs = 2, batch_size = 64)
+```
+
+## Beyond object-oriented development: the Functional API
+
+Was this example too much object-oriented development for you? You can also
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:928]:
```suggestion
If you prefer a less object-oriented way of programming, you can also
```

**Review by skeydan — APPROVED** [review]
Wow! Great new capabilities, and a super useful vignette :-)
Added a few things to ponder ... :-))

File: NEWS.md [hunk 18]
```diff
@@ -13,20 +13,24 @@ New Features:
 
 - The `keras` python module is exported
 
-- Introduced `%py_class%`, a new python class generator constructor.
+- Introduced `%py_class%`, a python class constructor constructor.
```
Comment (skeydan) [inline NEWS.md:16]: I think the double `constructor` here is intentional, but it's hard to understand... Does this make sense: 

> an R-language constructor for Python classes

? Or similar?

File: NEWS.md [hunk 19]
```diff
@@ -13,20 +13,24 @@ New Features:
...
   - Subclassing python classes: How to use `%py_class%`.
+  - Making new layers and models via subclassing.
   - Customizing what happens in fit (example of how to define a model, like a GAN, with a custom train step).
   - Writing your own callbacks.
 
 - All optimizer had argument `lr` renamed to `learning_rate`.
   (backwards compatibility is preserved, an R warning is now issued).
 
-- keras now imports the glue package
+- `Layer()` custom layer constructor is now lazy about initializing the python session and safe to use in the top level of an R package.
```
Comment (skeydan) [inline NEWS.md:27]:
```suggestion
- `Layer()` custom layer constructor is now lazy about initializing the Python session and safe to use on the top level of an R package.
```

File: NEWS.md [hunk 20]
```diff
@@ -58,12 +62,44 @@ New Features:
...
-- keras_model_sequential() gains the ability to accept arguments that
+  - New builtin loss functions:
+
+      -  `loss_huber()`
+      -  `loss_kl_divergence()`
+
+- Metric functions:
+  - All the metric functions gain the ability to return a `keras$metrics$Metric` instance if called without `y_true` and `y_pred`
+  - Each metric function is now documented separately, with a common `?Metric` topic demonstrating example usage.
+  - New built in metrics:
```
Comment (skeydan) [inline NEWS.md:76]:
```suggestion
  - New built-in metrics:
```

File: R/layer-custom.R [hunk 21]
```diff
@@ -117,3 +117,122 @@ compat_custom_KerasLayer_handler <- function(layer_class, args) {
...
+  }
+  args
+}
+
+
+
+
+#' Create a Keras Layer wrapper
+#'
+#' @param LayerClass A R6 or python class generator that inherits from
```
Comment (skeydan) [inline R/layer-custom.R:172]:
```suggestion
#' @param LayerClass A R6 or Python class generator that inherits from
```

Comment (skeydan) [reply R/layer-custom.R:172]: could you do the other `python`s by find-and-replace? :-)))

File: R/layer-custom.R [hunk 22]
```diff
@@ -117,3 +117,122 @@ compat_custom_KerasLayer_handler <- function(layer_class, args) {
...
+#'   `keras$layers$Layer`
+#' @param modifiers A named list of functions to modify to user-supplied
+#'   arguments before they are passed on to the class constructor. (e.g.,
+#'   `list(units = as.integer)`)
+#' @param convert Boolean, whether the python class and it's methods should by
```
Comment (skeydan) [inline R/layer-custom.R:177]:
```suggestion
#' @param convert Boolean, whether the Python class and its methods should by
```

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 23]
```diff
@@ -0,0 +1,1007 @@
...
+  layer_linear(4, 2)
+
+model(k_ones(c(2, 2)))
+
+model
+```
+
+Because the pattern above is so common, there is a convenience function that
+combines the steps of subclassing `keras$layers$Layer` and calling
+`create_layer_wrapper` on the output: the `Layer` function. The `layer_linear`
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:102]: I like it!

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 24]
```diff
@@ -0,0 +1,1007 @@
...
+                          trainable = TRUE)
+  },
+
+  call = function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+)
+```
+
+For the remainder of this vignette we'll be using the `%py_class%` constructor,
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:126]: not sure if these reasons will be convincing to readers ... I guess I'd rather remove that :-)

File: vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd [hunk 25]
```diff
@@ -0,0 +1,1007 @@
...
+```
+
+For more information, make sure to read the [Functional API guide](/guides/functional_api/).
+
+
+## Defining custom layers and models in an R package
+
+Unfortunately you can't use anything that creates references to python objects (e.g., `%py_class%`), at the top-level of an R package.
+
+When you build an R package, all the R files in the `R/` directory get sourced
```
Comment (skeydan) [inline vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd:943]: great explanation!

**Review by skeydan — APPROVED** [review]
Totally like the new features and explanations @t-kalinowski !
